core/server/worker/src/test/java/alluxio/worker/grpc/ShortCircuitBlockReadHandlerTest.java

```diff
@@ -0,0 +1,311 @@
+          .build(),
+      Configuration.modifiableGlobal()
+  );
+
+  @Rule
+  public TemporaryFolder mFolder = new TemporaryFolder();
+
+  // each tier's storage directory
+  private File mTier0Dir;
+  private File mTier1Dir;
+
+  // mocked response observer to interact with the handler
+  private TestResponseObserver mResponseObserver;
+
+  private final Closer mCloser = Closer.create();
+  // local store that backs the short circuit read
+  private BlockStore mBlockStore;
+  private ShortCircuitBlockReadHandler mTestHandler;
+
+  @Before
+  public void before() throws Exception {
+    // set up storage tier directories
+    mTier0Dir = mFolder.newFolder();
+    mTier1Dir = mFolder.newFolder();
+    mConfiguration.set(PATH_TEMPLATE.format(0), mTier0Dir.getAbsolutePath());
+    mConfiguration.set(PATH_TEMPLATE.format(1), mTier1Dir.getAbsolutePath());
+
+    // set up response observer
+    mResponseObserver = new TestResponseObserver();
+
+    // set up local storage
+    // todo(yangchen): BlockStore should not worry about the logic of reporting to master,
+    // could leave that work to BlockWorker so that we can get rid of these irrelevant mocks
+    BlockMasterClientPool pool = mock(BlockMasterClientPool.class);
+    BlockMasterClient cli = mock(BlockMasterClient.class);
+    doReturn(cli).when(pool).createNewResource();
+    doReturn(cli).when(pool).acquire();
+    mBlockStore = new MonoBlockStore(new TieredBlockStore(), pool,
+        mock(UfsManager.class), new AtomicReference<>(1L));
+
+    mTestHandler = new ShortCircuitBlockReadHandler(mBlockStore, mResponseObserver);
+  }
+
+  @Test
+  public void accessNonExistentBlock() {
+    // we started with empty directories without any block
+    long blockId = 1L;
+    OpenLocalBlockRequest request = createRequest(blockId, false);
+    mTestHandler.onNext(request);
+
+    // check that we get a proper error and no response
+    assertTrue(mResponseObserver.getResponses().isEmpty());
+    assertFalse(mResponseObserver.isCompleted());
+    Throwable t = mResponseObserver.getError();
+    assertTrue(t != null && t.getMessage().contains("BlockMeta not found"));
+  }
+
+  @Test
+  public void accessBlockNoPromote() throws Exception {
+    accessBlock(false);
+  }
+
+  @Test
+  public void accessBlockPromote() throws Exception {
+    accessBlock(true);
+  }
+
+  @Test
+  @Ignore("This test currently takes too long")
+  public void accessBlockPinned() throws Exception {
+    long blockId = 2L;
+    createLocalBlock(
+        1L, blockId, 0);
+
+    OpenLocalBlockRequest request = createRequest(blockId, false);
+    mTestHandler.onNext(request);
+
+    // block 2 should be pinned now
+    // and an attempt to remove the block from local store should fail
+    //todo(yangchen): TieredBlockStore's REMOVE_BLOCK_TIMEOUT_MS is too long for testing
+    assertThrows(Exception.class, () -> mBlockStore.removeBlock(3L, blockId));
+  }
+
+  @Test(timeout = 5000)
```

Comment:
What's taking so long? Can we tweak some config to make the delay shorter?

Reply:
The timeout acts as a safety net in case the block is not un-pinned because of the error. In that case the `TieredBlockStore#removeBlock` will block acquiring a write lock. Currently its timeout is 60_000ms.

Reply:
This is fixed together with the other review https://github.com/Alluxio/alluxio/pull/15710#discussion_r907523415